### PR TITLE
Use a structured error rather than an Aggregate error in discovery

### PR DESCRIPTION
**Pull request by smarterclayton**
Returning an aggregate hides errors from the cluster and results in the cause of failures being hard to determine.  Returning a single structured error with better messages will improve administrative debugging in prod environments.

Not required for 1.4, but would like this in 1.4.x.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32182)

### CODE REVIEW

File: pkg/client/typed/discovery/discovery_client.go [hunk 1]
```diff
@@ -203,7 +229,10 @@ func (d *DiscoveryClient) serverPreferredResources(namespaced bool) ([]unversion
 			results = append(results, groupVersion.WithResource(apiResource.Name))
 		}
 	}
-	return results, utilerrors.NewAggregate(allErrs)
+	if len(failedGroups) > 0 {
+		return results, &ErrGroupDiscoveryFailed{Groups: failedGroups}
```
Comment (soltysh) [inline pkg/client/typed/discovery/discovery_client.go]: Can you add a test case to cover this flow. While doing so please rename `client_test.go` to appropriate `discovery_client_test.go`.

**Comment by smarterclayton** [comment]
Test case added

**Comment by soltysh** [comment]
LGTM

**Comment by smarterclayton** [comment]
Test case added and renamed

**Comment by derekwaynecarr** [comment]
LGTM  - i want to bump priority as its flake related and get this in...

**Comment by pwittrock** [comment]
@derekparker @smarterclayton 

1.  Please provide a risk assessment in the description

2. How critical is this?  blocking / important / best effort?  We have dozens of cherrypick requests and each one creates a lot of overhead because large batches introduce a high-risk of destabilizing the tests in a way that is hard to track down.   We have to break them up into smaller batches and monitor the tests to catch issues early on and track them back to their root cause.

**Comment by smarterclayton** [comment]
Not required for 1.4 since we uncovered the cause of the failure.  I would like this in next point release though because it conceals errors from end users in production environments in surprising ways.

**Comment by pwittrock** [comment]
@smarterclayton Thanks.  I can revisit this later if we are ahead of schedule.

**Pull request merged**